pkg/objstore/prefixed_bucket.go

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package objstore
+
+import (
+	"context"
+	"io"
+	"strings"
+)
+
+type PrefixedBucket struct {
+	bkt    Bucket
+	prefix string
+}
+
+func NewPrefixedBucket(bkt Bucket, prefix string) Bucket {
+	if validPrefix(prefix) {
+		return &PrefixedBucket{bkt: bkt, prefix: strings.Trim(prefix, DirDelim)}
+	}
+
+	return bkt
+}
+
+func validPrefix(prefix string) bool {
+	prefix = strings.Replace(prefix, "/", "", -1)
+	return len(prefix) > 0
+}
+
+func conditionalPrefix(prefix, name string) string {
+	if len(name) > 0 && len(prefix) > 0 {
+		return prefix + DirDelim + name
```

Comment:
nit: let's call `withPrefix` here? actually I think we only need to check for `len(name) > 0` here because `len(prefix) > 0` is already given our initial checks?

Reply:
Makes sense. Done!